### PR TITLE
Fix querying on nested has-many associations

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: lib/scoped_search/query_builder.rb [hunk 1]
```diff
@@ -269,39 +269,43 @@ def find_has_many_through_association(field, through)
...
+      join_reflections.zip(table_names).reduce(sql) do |acc, (reflection, previous_table)|
+        klass = reflection.method(:join_keys).arity == 1 ? [reflection.klass] : []
+        fk1, pk1 = reflection.join_keys(*klass).values # We are joining the tables "in reverse", so the PK and FK are swapped
+        t1 = previous_table
+
+        t2 = reflection.table_name
+
+        # primary and foreign keys + optional conditions for the joins
+        condition_many_to_middle = if with_polymorphism?(many_class, field.klass, reflection.name, reflection.klass)
+                                     field.reflection_conditions(definition.reflection_by_name(field.klass, many_class.table_name))
```
Comment (adamruzicka) [inline lib/scoped_search/query_builder.rb:299]: This is nonsense for polymorphic has-many-through associations, looking into it

**Comment by ares** [comment]
More background or test explaining the issue would be great once you're ready :-)

**Comment by parthaa** [comment]
I ran a couple of tests with this PR. Not sure it works.
Query before https://github.com/wvanbergen/scoped_search/pull/191 


```Frame number: 0/30
[1] pry(main)> Host.search_for('applicable_errata_issued = "10 days ago"')
2020-07-14T13:01:30 [D|sql|]   SQL (5.0ms)  
=> []
```
Check the generated SQL query => https://gist.github.com/parthaa/9ddcb7d80fa4c38a40fe95be88617e79


With the changes in the PR the generated queries for the same call look like
https://gist.github.com/parthaa/32edd0a1882f15bfcb9da6c69b39e05f
and
https://gist.github.com/parthaa/2b66a0be71ddce8070a87fad3b9f27a9

The output was slightly different from pre 191 PR (that would return a [ ] .) This one returns a relation

```
[1] pry(main)> Host.search_for('applicable_errata_issued = "10 days ago"')
=> #<Host::Managed::ActiveRecord_Relation:0x5f8bccc>
```
more over
``` Host.search_for('applicable_errata_issued = "10 days ago"').count``` fails with
```
[2] pry(main)> Host.search_for('applicable_errata_issued = "10 days ago"').count
2020-07-14T17:15:36 [D|sql|]    (3.1ms)  SELECT COUNT(DISTINCT "hosts"."id") FROM "hosts" LEFT OUTER JOIN "katello_content_facets" ON "katello_content_facets"."host_id" = "hosts"."id" LEFT OUTER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."content_facet_id" = "katello_content_facets"."id" LEFT OUTER JOIN "katello_errata" ON "katello_errata"."id" = "katello_content_facet_errata"."erratum_id" WHERE "hosts"."type" = $1 AND (("hosts"."id" IN (SELECT 
"hosts"."id" FROM "hosts"          INNER JOIN "katello_content_facets"
 |           ON "hosts"."id" = "katello_content_facets"."host_id" 
 |           INNER JOIN "katello_errata"
 |           ON "katello_content_facets"."erratum_id" = "katello_errata"."id" 
 |  WHERE "katello_errata"."issued" = '2020-07-04' )))  [["type", "Host::Managed"]]
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column katello_content_facets.erratum_id does not exist
LINE 4:           ON "katello_content_facets"."erratum_id" = "katell...
                     ^

from /home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:675:in `exec_params'
Caused by PG::UndefinedColumn: ERROR:  column katello_content_facets.erratum_id does not exist
LINE 4:           ON "katello_content_facets"."erratum_id" = "katell...
                     ^

from /home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:675:in `exec_params'

```

**Comment by adamruzicka** [comment]
In order for this to work, the has-something-through chain needs to be *fully* defined all the way through. In Katello there was a missing step (the association went straight from content facet to errata without mentioning the join table between the two), see https://github.com/Katello/katello/pull/8839.

**Comment by parthaa** [comment]
> In order for this to work, the has-something-through chain needs to be _fully_ defined all the way through. In Katello there was a missing step (the association went straight from content facet to errata without mentioning the join table between the two), see [Katello/katello#8839](https://github.com/Katello/katello/pull/8839).

Hmm even after applying the katello change I got a similar respone
```
[2] pry(main)> Host.search_for('applicable_errata_issued = "10 days ago"').pluck(:id)
2020-07-15T19:40:49 [D|sql|]    (3.1ms)  SELECT "hosts"."id" FROM "hosts" LEFT OUTER JOIN "katello_content_facets" ON "katello_content_facets"."host_id" = "hosts"."id" LEFT OUTER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."content_facet_id" = "katello_content_facets"."id" LEFT OUTER JOIN "katello_errata" ON "katello_errata"."id" = "katello_content_facet_errata"."erratum_id" WHERE "hosts"."type" = $1 AND (("hosts"."id" IN (SELECT "hosts"."id" FROM "hosts"          INNER JOIN "katello_content_facets"
 |           ON "hosts"."id" = "katello_content_facets"."host_id" 
 |           INNER JOIN "katello_errata"
 |           ON "katello_content_facets"."erratum_id" = "katello_errata"."id" 
 |  WHERE "katello_errata"."issued" = '2020-07-05' ))) ORDER BY "hosts"."name" ASC  [["type", "Host::Managed"]]
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column katello_content_facets.erratum_id does not exist
LINE 4:           ON "katello_content_facets"."erratum_id" = "katell...
                     ^

from /home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:675:in `exec_params'
Caused by PG::UndefinedColumn: ERROR:  column katello_content_facets.erratum_id does not exist
LINE 4:           ON "katello_content_facets"."erratum_id" = "katell...
                     ^

from /home/vagrant/foreman/.vendor/ruby/2.5.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql_adapter.rb:675:in `exec_params'
```

**Comment by adamruzicka** [comment]
I have the katello patch applied on top of katello-3.16.0.rc3 together with this branch in scoped search and the tables seem to get joined correctly. Is it possible that you haven't restarted something after applying the patches?

```
[1] pry(main)> Host.search_for('applicable_errata_issued = "10 days ago"').pluck(:id)
=> []

[2] pry(main)> puts Host.search_for('applicable_errata_issued = "10 days ago"').to_sql; nil
SELECT "hosts"."id" AS t0_r0, "hosts"."name" AS t0_r1, "hosts"."last_compile" AS t0_r2, "hosts"."last_report" AS t0_r3, "hosts"."updated_at" AS t0_r4, "hosts"."created_at" AS t0_r5, "hosts"."root_pass" AS t0_r6, "hosts"."architecture_id" AS t0_r7, "hosts"."operatingsystem_id" AS t0_r8, "hosts"."environment_id" AS t0_r9, "hosts"."ptable_id" AS t0_r10, "hosts"."medium_id" AS t0_r11, "hosts"."build" AS t0_r12, "hosts"."comment" AS t0_r13, "hosts"."disk" AS t0_r14, "hosts"."installed_at" AS t0_r15, "hosts"."model_id" AS t0_r16, "hosts"."hostgroup_id" AS t0_r17, "hosts"."owner_id" AS t0_r18, "hosts"."owner_type" AS t0_r19, "hosts"."enabled" AS t0_r20, "hosts"."puppet_ca_proxy_id" AS t0_r21, "hosts"."managed" AS t0_r22, "hosts"."use_image" AS t0_r23, "hosts"."image_file" AS t0_r24, "hosts"."uuid" AS t0_r25, "hosts"."compute_resource_id" AS t0_r26, "hosts"."puppet_proxy_id" AS t0_r27, "hosts"."certname" AS t0_r28, "hosts"."image_id" AS t0_r29, "hosts"."organization_id" AS t0_r30, "hosts"."location_id" AS t0_r31, "hosts"."type" AS t0_r32, "hosts"."otp" AS t0_r33, "hosts"."realm_id" AS t0_r34, "hosts"."compute_profile_id" AS t0_r35, "hosts"."provision_method" AS t0_r36, "hosts"."grub_pass" AS t0_r37, "hosts"."discovery_rule_id" AS t0_r38, "hosts"."global_status" AS t0_r39, "hosts"."lookup_value_matcher" AS t0_r40, "hosts"."openscap_proxy_id" AS t0_r41, "hosts"."pxe_loader" AS t0_r42, "hosts"."initiated_at" AS t0_r43, "hosts"."build_errors" AS t0_r44, "katello_errata"."id" AS t1_r0, "katello_errata"."pulp_id" AS t1_r1, "katello_errata"."errata_id" AS t1_r2, "katello_errata"."created_at" AS t1_r3, "katello_errata"."updated_at" AS t1_r4, "katello_errata"."issued" AS t1_r5, "katello_errata"."updated" AS t1_r6, "katello_errata"."errata_type" AS t1_r7, "katello_errata"."severity" AS t1_r8, "katello_errata"."title" AS t1_r9, "katello_errata"."solution" AS t1_r10, "katello_errata"."description" AS t1_r11, "katello_errata"."summary" AS t1_r12, "katello_errata"."reboot_suggested" AS t1_r13, "katello_errata"."migrated_pulp3_href" AS t1_r14 FROM "hosts" LEFT OUTER JOIN "katello_content_facets" ON "katello_content_facets"."host_id" = "hosts"."id" LEFT OUTER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."content_facet_id" = "katello_content_facets"."id" LEFT OUTER JOIN "katello_errata" ON "katello_errata"."id" = "katello_content_facet_errata"."erratum_id" WHERE "hosts"."type" = 'Host::Managed' AND (("hosts"."id" IN (SELECT "hosts"."id" FROM "hosts"          INNER JOIN "katello_content_facets"
          ON "hosts"."id" = "katello_content_facets"."host_id" 
          INNER JOIN "katello_content_facet_errata"
          ON "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" 
          INNER JOIN "katello_errata"
          ON "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" 
 WHERE "katello_errata"."issued" = '2020-07-06' ))) ORDER BY "hosts"."name" ASC
=> nil
```

**Comment by parthaa** [comment]
Ok you are correct!. It was some cached thing in my development environment causing this. ACK from my perspective
```ruby
[13] pry(main)> Host.search_for('applicable_errata_issued < "10 days ago"')
D, [2020-07-31T15:36:46.539568 #2243] DEBUG -- :   SQL (11.0ms)  SELECT "hosts"."id" AS t0_r0, "hosts"."name" AS t0_r1, "hosts"."last_compile" AS t0_r2, "hosts"."last_report" AS t0_r3, "hosts"."updated_at" AS t0_r4, "hosts"."created_at" AS t0_r5, "hosts"."root_pass" AS t0_r6, "hosts"."architecture_id" AS t0_r7, "hosts"."operatingsystem_id" AS t0_r8, "hosts"."environment_id" AS t0_r9, "hosts"."ptable_id" AS t0_r10, "hosts"."medium_id" AS t0_r11, "hosts"."build" AS t0_r12, "hosts"."comment" AS t0_r13, "hosts"."disk" AS t0_r14, "hosts"."installed_at" AS t0_r15, "hosts"."model_id" AS t0_r16, "hosts"."hostgroup_id" AS t0_r17, "hosts"."owner_id" AS t0_r18, "hosts"."owner_type" AS t0_r19, "hosts"."enabled" AS t0_r20, "hosts"."puppet_ca_proxy_id" AS t0_r21, "hosts"."managed" AS t0_r22, "hosts"."use_image" AS t0_r23, "hosts"."image_file" AS t0_r24, "hosts"."uuid" AS t0_r25, "hosts"."compute_resource_id" AS t0_r26, "hosts"."puppet_proxy_id" AS t0_r27, "hosts"."certname" AS t0_r28, "hosts"."image_id" AS t0_r29, "hosts"."organization_id" AS t0_r30, "hosts"."location_id" AS t0_r31, "hosts"."type" AS t0_r32, "hosts"."otp" AS t0_r33, "hosts"."realm_id" AS t0_r34, "hosts"."compute_profile_id" AS t0_r35, "hosts"."provision_method" AS t0_r36, "hosts"."grub_pass" AS t0_r37, "hosts"."global_status" AS t0_r38, "hosts"."lookup_value_matcher" AS t0_r39, "hosts"."pxe_loader" AS t0_r40, "hosts"."initiated_at" AS t0_r41, "hosts"."build_errors" AS t0_r42, "katello_errata"."id" AS t1_r0, "katello_errata"."pulp_id" AS t1_r1, "katello_errata"."errata_id" AS t1_r2, "katello_errata"."created_at" AS t1_r3, "katello_errata"."updated_at" AS t1_r4, "katello_errata"."issued" AS t1_r5, "katello_errata"."updated" AS t1_r6, "katello_errata"."errata_type" AS t1_r7, "katello_errata"."severity" AS t1_r8, "katello_errata"."title" AS t1_r9, "katello_errata"."solution" AS t1_r10, "katello_errata"."description" AS t1_r11, "katello_errata"."summary" AS t1_r12, "katello_errata"."reboot_suggested" AS t1_r13, "katello_errata"."migrated_pulp3_href" AS t1_r14 FROM "hosts" LEFT OUTER JOIN "katello_content_facets" ON "katello_content_facets"."host_id" = "hosts"."id" LEFT OUTER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."content_facet_id" = "katello_content_facets"."id" LEFT OUTER JOIN "katello_errata" ON "katello_errata"."id" = "katello_content_facet_errata"."erratum_id" WHERE "hosts"."type" = $1 AND (("hosts"."id" IN (SELECT "hosts"."id" FROM "hosts"          INNER JOIN "katello_content_facets"
          ON "hosts"."id" = "katello_content_facets"."host_id" 
          INNER JOIN "katello_content_facet_errata"
          ON "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" 
          INNER JOIN "katello_errata"
          ON "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" 
 WHERE "katello_errata"."issued" < '2020-07-21' ))) ORDER BY "hosts"."name" ASC  [["type", "Host::Managed"]]
=> [#<Host::Managed:0x00000000128e55f0
  id: 3,
  name: "centos7.sparta.example.com",
  last_compile: Fri, 24 Jul 2020 13:57:05 UTC +00:00,
  last_report: nil,
  updated_at: Fri, 24 Jul 2020 13:57:07 UTC +00:00,
  created_at: Fri, 24 Jul 2020 13:51:02 UTC +00:00,
  root_pass: nil,
  architecture_id: 1,
  operatingsystem_id: 1,
  environment_id: nil,
  ptable_id: nil,
  medium_id: nil,
  build: false,
  comment: nil,
  disk: nil,
  installed_at: nil,
  model_id: 1,
  hostgroup_id: nil,
  owner_id: 1,
  owner_type: "User",
  enabled: true,
  puppet_ca_proxy_id: nil,
  managed: false,
  use_image: nil,
  image_file: nil,
  uuid: nil,
  compute_resource_id: nil,
  puppet_proxy_id: nil,
  certname: nil,
  image_id: nil,
  organization_id: 1,
  location_id: 2,
  type: "Host::Managed",
  otp: nil,
  realm_id: nil,
  compute_profile_id: nil,
  provision_method: nil,
  grub_pass: "",
  global_status: 1,
  lookup_value_matcher: [FILTERED],
  pxe_loader: nil,
  initiated_at: nil,
  build_errors: nil>]
[14] pry(main)> 



```

**Comment by ares** [comment]
Thanks @adamruzicka, merging!

**Pull request merged**